### ISSUE TITLE
Allow status items with error background color

### ISSUE THREAD

**Issue by dbaeumer** [microsoft/vscode#110214]
Using the theme color for status bar items might not meet accessibility requirements (contrast). Is there another color I should use or do we actually need another theme color for this.
<img width="274" alt="capture" src="https://user-images.githubusercontent.com/1931590/98530704-958a6c80-227f-11eb-91c2-7c4e61d1026d.png">

**Comment by gjsjohnmurray** [comment]
related #108024

**Comment by misolori** [comment]
@aeschli I'm wondering if we need to introduce something like `statusBar.errorForeground` in order to allow custom error colors in the status bar?

**Comment by misolori** [comment]
Started working on this and it appears that other themes seem to work pretty well for showing errors in the status bar except for our default Dark/Light theme, because we use a bright blue:

![Kapture 2020-11-23 at 11 51 46](https://user-images.githubusercontent.com/35271042/100008421-4e71a080-2d82-11eb-84b1-7b7544a1959f.gif)

![image](https://user-images.githubusercontent.com/35271042/100008315-1ec29880-2d82-11eb-9156-c0f2e13d3c3a.png)

Currently, the brightest color we can use and still pas color contrast ratio is white:

<img width="300" src="https://user-images.githubusercontent.com/35271042/99990406-9e436e00-2d68-11eb-9d1b-2d92c61671ec.png">

![image](https://user-images.githubusercontent.com/35271042/100008951-f9825a00-2d82-11eb-9f79-f8a4447a9195.png)

This means that we wouldn't be able to highlight any text in the status bar in our default Dark/Light themes and would need to set the new color tokens to white. Should we still add these?

**Comment by misolori** [comment]
The other alternative is to add a background color to the status bar item, which is what https://github.com/microsoft/vscode/issues/28739 covers and would need an update to the API:

![image](https://user-images.githubusercontent.com/35271042/100017031-26d50500-2d8f-11eb-9f61-7bb072b91685.png)

**Comment by usernamehw** [comment]
The api looks like a better choice and it's an easy one to code. Users can hide items with right click if a random extension would abuse it and it's already possible to create an ugly prominent status bar items with a bunch of emojis `🟥🔴🔺🔻🟥🔴🍎🔴`.

Api is an easy win. (Unless there are some arguments against it).

**Comment by misolori** [comment]
After discussing this with the team, the best way to accomplish this would be to update the API and only allow for "error" status bar items to have this styling. We wouldn't want to allow for any bg colors to be added to avoid this getting abused, so we'd still introduce `statusBar.errorForeground` and `statusBar.errorBackground` as color tokens.

@aeschli can you work on getting this updated in the API and going through our API proposal flow? Thanks.

**Comment by aeschli** [comment]
Makes sense to me to have a 'error' property on the status bar entry.
@bpasero is the owner of the status bar... (I can help out if needed)

**Comment by bpasero** [comment]
@misolori @aeschli if I am to implement this I need the requirements straight and currently I am not entirely sure I get it, since I was also not part of any API discussion.

Today we have `color` option:

https://github.com/microsoft/vscode/blob/c6e908997b91e92b7287eba2504d0b50a27a05ba/src/vs/vscode.d.ts#L5376-L5376

Are we saying we add something like `isError` or rather `backgroundColor` option? If we later want to revisit this, would it make sense to allow to only set one color to `backgroundColor`, namely the error color?

**Comment by usernamehw** [comment]
Would it be possible to add background color api with a option for user to opt out of any api background in status bar?

```js
"workbench.statusBar.allowBackgroundColor": true,
```

**Comment by misolori** [comment]
@bpasero well, we'd need to have some way to control the styling for themes because the color might not always work out of the box, so we'd still want to introduce color tokens for  `statusBar.errorForeground` and `statusBar.errorBackground`. Would it then make sense to go down the `isError` route if we need these additional tokens?

**Comment by bpasero** [comment]
From UX meeting, also looking bad: <img width="225" alt="Screenshot 2020-11-25 at 17 32 40" src="https://user-images.githubusercontent.com/900690/100255799-495c4f00-2f44-11eb-9ed1-cdfbd5c09b84.png">

**Comment by misolori** [comment]
Yea, that's why adding a colored background color would be necessary to make this look better:

![image](https://user-images.githubusercontent.com/35271042/100256862-24db8500-2efa-11eb-831b-7fba78141848.png)

**Comment by bpasero** [comment]
@misolori @aeschli so then would we ignore the `color` API if something like `isError` is set to true?

Somehow I feel such an API is not cutting it and we will come back adding more options for more colors. That is why I thought having `backgroundColor` API with a reduced set of allowed colors would provide us with more options. 

I don't know how in our API we surface theme colors though, maybe @aeschli knows.

**Comment by aeschli** [comment]
I agree with @bpasero that 'isError' might just be the beginning and we see other requests like `isWarning`.
I like the idea of a statusbar.backgoundColor API with a reduced set of allowed colors.

Because ThemeColor is a class, I don't think we can enforce it in the API, but will have to spec it. Could also add a namespace with a list of available ThemeColors.

```ts
{
   /** 
    * Background color of the status bar item. 
    * The supported colors are limited to ThemeColors with 
    * id `statusBar.errrorBackground`. Other ThemeColors will be ignored. 
    */
   backgoundColor: ThemeColor | undefined 
} 
```


```ts
{
   export namespace StatusBarBackgroundColors {
      /** Color of error backgrounds for status bar entries */
      export const error: ThemeColor;
   }
   /**
    *  Background color of the status bar item. The supported colors are 
    * limited to colors listed in StatusBarBackgroundColors
    */
   backgroundColor: ThemeColor | undefined 
} 
```

**Comment by bpasero** [comment]
I like the first proposal 👍 . Would `color` still be applied then?

**Comment by aeschli** [comment]
Yes, `color` is still applied and in fact users are advised to set it to `statusBar.errrorForeground`
```ts
{
   /** 
    * Background color of the status bar item. 
    * The supported colors are limited to ThemeColors with 
    * id `statusBar.errrorBackground`. Other ThemeColors will be ignored. 
    * When setting the background color to `statusBar.errrorBackground`, it is recommended to 
    * also set `color` to `statusBar.errrorForeground`
    */
   backgoundColor: ThemeColor | undefined 
} 
```

**Comment by bpasero** [comment]
PR open: https://github.com/microsoft/vscode/pull/111353

@misolori would you be able to add the relevant `statusBar.errorBackground` and `statusBar.errorForeground` to our themes? Feel free to push to that branch directly.

**Comment by bpasero** [comment]
Actually the name should be `statusBarItem.errorBackground` and `statusBarItem.errorForeground` to align with existing color names.

**Comment by jrieken** [comment]
@bpasero make sure to keep this issue around & open until finalised

**Comment by misolori** [comment]
I've updated the **default** color tokens to use a darker color of `errorForeground` since majority of themes have a much lighter color and to pass color contrast ratio in our dark/light themes:

![ezgif com-gif-maker (21)](https://user-images.githubusercontent.com/35271042/100672300-952d4080-3316-11eb-910d-7000856927ad.gif)

**Comment by bpasero** [comment]
Let's try to stabilize this for January 2021.

**Comment by bpasero** [comment]
Verification: you can use the `backgroundColor` API on a `StatusBarItem` without having to enable proposed API.

**Comment by jrieken** [comment]
@bpasero is this on-testplan or verification needed?

**Comment by bpasero** [comment]
@jrieken yeah it was "on testplan" last milestone to test the API, but now we need a verification that the API can be used without proposed API flag. That is what I left the verification in this issue for.

**Comment by lszomoru** [comment]
@jrieken, are you working on this one? If not, I am happy to verify it.

**Comment by meganrogge** [comment]
I used the status bar extension sample to try to verify this, adding these lines of code within extensions.ts

```
function updateStatusBarItem(): void {
	const n = getNumberOfSelectedLines(vscode.window.activeTextEditor);
	if (n > 0) {
		myStatusBarItem.text = `$(megaphone) ${n} line(s) selected`;
		myStatusBarItem.backgroundColor = new vscode.ThemeColor('statusBar.errorBackground');
		myStatusBarItem.color = new vscode.ThemeColor('statusBar.errorForeground');
		myStatusBarItem.show();
	} else {
		myStatusBarItem.hide();
	}
}
```

It appears that the text is actually invisible now and the error color isn't present. What am I doing wrong?

![recording (10)](https://user-images.githubusercontent.com/29464607/106038665-168e1e80-608d-11eb-9116-a5f610d0f1a6.gif)

**Comment by meganrogge** [comment]
@misolori realized it's` statusBarItem.errorBackground` and `statusBarItem.foregroundBackground` and we verified those work.

The above gif demos the bug that happens when those aren't set correctly though.

**Comment by bpasero** [comment]
Good catch, I reported as https://github.com/microsoft/vscode/issues/115292
